crates/circuit/src/dag_circuit.rs

```diff
@@ -6437,4 +6447,28 @@ impl DAGCircuit {
     }
 }
 
+/// Add to global phase. Global phase can only be Float or ParameterExpression so this
+/// does not handle the full possibility of parameter values.
+fn add_global_phase(py: Python, phase: &Param, other: &Param) -> PyResult<Param> {
+    Ok(match [phase, other] {
+        [Param::Float(a), Param::Float(b)] => Param::Float(a + b),
+        [Param::Float(a), Param::ParameterExpression(b)] => Param::ParameterExpression(
+            b.clone_ref(py)
+                .call_method1(py, intern!(py, "__radd__"), (*a,))?,
+        ),
+        [Param::ParameterExpression(a), Param::Float(b)] => Param::ParameterExpression(
+            a.clone_ref(py)
+                .call_method1(py, intern!(py, "__add__"), (*b,))?,
+        ),
+        [Param::ParameterExpression(a), Param::ParameterExpression(b)] => {
+            Param::ParameterExpression(a.clone_ref(py).call_method1(
+                py,
+                intern!(py, "__add__"),
+                (b,),
+            )?)
+        }
+        _ => panic!("Invalid global phase"),
```

Comment:
At the point that we're propagating a `PyResult` now, we probably ought to return a `PyTypeError` in this case rather than a panic?